pkg/istio/onboarding/controller.go

```diff
@@ -0,0 +1,246 @@
+	config := model.Config{
+		ConfigMeta: model.ConfigMeta{
+			Type:    model.ClusterRbacConfig.Type,
+			Name:    model.DefaultRbacConfigName,
+			Group:   model.ClusterRbacConfig.Group + model.IstioAPIGroupDomain,
+			Version: model.ClusterRbacConfig.Version,
+		},
+		Spec: &v1alpha1.RbacConfig{
+			Mode: v1alpha1.RbacConfig_ON_WITH_INCLUSION,
+			Inclusion: &v1alpha1.RbacConfig_Target{
+				Services: services,
+			},
+		},
+	}
+
+	_, err := c.store.Create(config)
+	return err
+}
+
+// getServiceList extracts all services from the indexer with the authz
+// annotation set to true.
+func (c *Controller) getServiceList() []string {
+	cacheServiceList := c.serviceIndexInformer.GetIndexer().List()
+	serviceList := make([]string, 0)
+
+	for _, service := range cacheServiceList {
+		svc, ok := service.(*v1.Service)
+		if !ok {
+			log.Println("Could not cast to service object, skipping service list addition...")
+			continue
+		}
+
+		key, exists := svc.Annotations[authzEnabledAnnotation]
+		if exists && key == authzEnabled {
+			serviceName := svc.Name + "." + svc.Namespace + "." + c.dnsSuffix
+			serviceList = append(serviceList, serviceName)
+		}
+	}
+
+	return serviceList
+}
+
+// sync decides whether to create / update / delete the ClusterRbacConfig
+// object based on the current onboarded services in the cluster
+func (c *Controller) sync() error {
+	serviceList := c.getServiceList()
+	config := c.store.Get(model.ClusterRbacConfig.Type, model.DefaultRbacConfigName, "")
+	if config == nil && len(serviceList) == 0 {
+		log.Println("Service list is empty and cluster rbac config does not exist, skipping sync...")
+		return nil
+	}
+
+	if config == nil {
+		log.Println("Creating cluster rbac config...")
+		return c.createClusterRbacConfig(serviceList)
+	}
+
+	clusterRbacConfig, ok := config.Spec.(*v1alpha1.RbacConfig)
+	if !ok {
+		return errors.New("Could not cast to ClusterRbacConfig")
+	}
+
+	if clusterRbacConfig.Inclusion == nil {
+		return errors.New("ClusterRbacConfig inclusion service list is empty")
+	}
+
+	newServices := compareServiceLists(serviceList, clusterRbacConfig.Inclusion.Services)
+	c.addServices(newServices, clusterRbacConfig)
+
+	oldServices := compareServiceLists(clusterRbacConfig.Inclusion.Services, serviceList)
+	c.deleteServices(oldServices, clusterRbacConfig)
+
+	if len(clusterRbacConfig.Inclusion.Services) == 0 {
+		log.Println("Deleting cluster rbac config...")
+		return c.store.Delete(model.ClusterRbacConfig.Type, model.DefaultRbacConfigName, v1.NamespaceDefault)
+	}
+
+	if len(newServices) > 0 || len(oldServices) > 0 {
+		log.Println("Updating cluster rbac config...")
+		_, err := c.store.Update(model.Config{
+			ConfigMeta: config.ConfigMeta,
+			Spec:       clusterRbacConfig,
+		})
+		return err
+	}
+
+	log.Println("Sync state is current, no changes needed...")
+	return nil
+}
+
+// compareServices compares the two arrays and returns the difference
```

Comment:
can we update the comment to explicitly indicate that this func returns the items in first list that are not in the second list

Reply:
Updated!